xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -2022,43 +2005,67 @@ private void cleanUpResourceTimers() {
     }
   }
 
-  private void handleResourcesAccepted(
-      ResourceType type, Map<String, ParsedResource> parsedResources, String version,
-      String nonce) {
-    ackResponse(type, version, nonce);
-
+  private void handleResourceUpdate(
+      ResourceType type, Map<String, ParsedResource> parsedResources, Set<String> invalidResources,
+      Set<String> retainedResources, String version, String nonce, List<String> errors) {
+    String errorDetail = null;
+    if (errors.isEmpty()) {
+      checkArgument(invalidResources.isEmpty(), "found invalid resources but missing errors");
+      ackResponse(type, version, nonce);
+    } else {
+      errorDetail = Joiner.on('\n').join(errors);
+      getLogger().log(XdsLogLevel.WARNING,
+          "Failed processing {0} Response version {1} nonce {2}. Errors:\n{3}",
+          type, version, nonce, errorDetail);
+      nackResponse(type, nonce, errorDetail);
+    }
     long updateTime = timeProvider.currentTimeNanos();
     for (Map.Entry<String, ResourceSubscriber> entry : getSubscribedResourcesMap(type).entrySet()) {
       String resourceName = entry.getKey();
       ResourceSubscriber subscriber = entry.getValue();
+      // Attach error details to the subscribed resources that included in the ADS update.
+      if (invalidResources.contains(resourceName)) {
+        subscriber.onRejected(version, updateTime, errorDetail);
+      }
       // Notify the watchers.
       if (parsedResources.containsKey(resourceName)) {
         subscriber.onData(parsedResources.get(resourceName), version, updateTime);
       } else if (type == ResourceType.LDS || type == ResourceType.CDS) {
+        if (subscriber.data != null && invalidResources.contains(resourceName)) {
```

Comment:
It seems this condition should move into the `invalidResources.contains(resourceName)` case above and then we combine the `invalidResources.contains(resourceName` into the if-else chain.

Reply:
I'm not sure how to simplify this. The `subscriber.onAbsent()` below still need (the negation of) this condition. I grouped the logic regarding `retainedResources` and `onAbsent()` together because they are related.

Reply:
I was thinking:

```java
if (invalidResources.contains(resourceName)) {
  subscriber.onRejected(version, updateTime, errorDetail);
  if (subscriber.data != null) {
    if (type == ResourceType.LDS) {
      ...
      retainedResources.add(rdsName);
    } else if (type == ResourceType.CDS) {
      ...
      retainedResources.add(edsName);
    }
  }
} else if (parsedResources.containsKey(resourceName)) {
  subscriber.onData(parsedResources.get(resourceName), version, updateTime);
} else if (type == ResourceType.LDS || type == ResourceType.CDS) {
  subscriber.onAbsent();
}
```

But I can live with what's here.